### PR TITLE
refactor parallel check meta data, refactor TableMetaDataLoaderEngine.

### LINKED ISSUE #11658: Load metadata,ShardingTableMetaDataBuilder#parallelLoadTables performance waste.
### Is your feature request related to a problem?
not a problem, but a little performance waste about
### Describe the feature you would like.
sharding config, when i make `check-table-metadata-enabled: true`
then it'll go through `TableMetaDataBuilder#parallelLoadTables`
Then i found tableFutureMap will throw away some future, which is already load table.
If I want to check whether the tables of all data nodes are consistent, then I shouldn't just use the table name as the key.
If I just want to check all the tables on one node, then I shouldn't load all the nodes.
image

### PR DESCRIPTION
**Pull request by tuichenchuxin**
Fixes #11658.

Changes proposed in this pull request:
- refactor parallel check meta data
- refactor `TableMetaDataLoaderEngine`

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/util/TableMetaDataUtil.java [hunk 1]
```diff
@@ -55,4 +58,24 @@
...
     }
+    
+    /**
+     * Get data source all actual table groups.
+     *
+     * @param tableNames table name collection
+     * @param materials materials
+     * @return datasource and table collection map
+     */
+    public static Map<String, Collection<String>> getDataSourceAllActualTableGroups(final Collection<String> tableNames, final SchemaBuilderMaterials materials) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/util/TableMetaDataUtil.java:69]: @tuichenchuxin You may consider publishing a community task to add test cases for TableMetaDataUtil.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java [hunk 2]
```diff
@@ -46,6 +47,21 @@
...
+     * Load table meta data.
+     *
+     * @param dataSource data source
+     * @param tableName table name
+     * @param databaseType database type
+     * @return table meta data
+     * @throws SQLException SQL exception
+     */
+    public static Optional<TableMetaData> load(final DataSource dataSource, final String tableName, final DatabaseType databaseType) throws SQLException {
+        Map<String, Collection<String>> dataSourceTable = Collections.singletonMap("ds", Collections.singleton(tableName));
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java:60]: @tuichenchuxin Please don't hard code here.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java [hunk 3]
```diff
@@ -58,8 +74,14 @@
     public static Collection<TableMetaData> load(final Map<String, Collection<String>> dataSourceTable, final DatabaseType databaseType,
                                                  final Map<String, DataSource> dataSourceMap) throws SQLException {
         Optional<DialectTableMetaDataLoader> dialectTableMetaDataLoader = findDialectTableMetaDataLoader(databaseType);
-        return dialectTableMetaDataLoader.isPresent() ? loadByDialect(dialectTableMetaDataLoader.get(), dataSourceTable, dataSourceMap)
-                : loadByDefault(dataSourceTable, databaseType, dataSourceMap);
+        if (dialectTableMetaDataLoader.isPresent()) {
+            try {
+                return loadByDialect(dialectTableMetaDataLoader.get(), dataSourceTable, dataSourceMap);
+            } catch (SQLException e) {
+                return loadByDefault(dataSourceTable, databaseType, dataSourceMap);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java:81]: @tuichenchuxin Please print out the exception information and modify the exception variables.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoadMaterials.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+import javax.sql.DataSource;
+import java.util.Collection;
+
+/**
+ * TableMetaData load materials.
+ */
+@RequiredArgsConstructor
+@Getter
+public class TableMetaDataLoadMaterials {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoadMaterials.java:32]: @tuichenchuxin Please add final for this class and subclass.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java [hunk 5]
```diff
@@ -41,44 +41,47 @@
 import java.util.concurrent.TimeUnit;
 
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Slf4j
 public final class TableMetaDataLoaderEngine {
 
     private static final ExecutorService EXECUTOR_SERVICE = new ThreadPoolExecutor(Runtime.getRuntime().availableProcessors() * 2, Runtime.getRuntime().availableProcessors() * 2,
             0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>(), new ThreadFactoryBuilder().setDaemon(true).setNameFormat("ShardingSphere-TableMetaDataLoaderEngine-%d").build());
-
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java:48]: @tuichenchuxin Please keep this space.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java [hunk 6]
```diff
@@ -41,44 +41,47 @@
...
-        Optional<DialectTableMetaDataLoader> dialectTableMetaDataLoader = findDialectTableMetaDataLoader(databaseType);
-        return dialectTableMetaDataLoader.isPresent() ? loadByDialect(dialectTableMetaDataLoader.get(), dataSourceTable, dataSourceMap)
-                : loadByDefault(dataSourceTable, databaseType, dataSourceMap);
+    public static Collection<TableMetaData> load(final TableMetaDataLoadMaterials materials) throws SQLException {
+        Optional<DialectTableMetaDataLoader> dialectTableMetaDataLoader = findDialectTableMetaDataLoader(materials.getDatabaseType());
+        if (dialectTableMetaDataLoader.isPresent()) {
+            try {
+                return loadByDialect(dialectTableMetaDataLoader.get(), materials);
+            } catch (final SQLException | ShardingSphereException ex) {
+                log.warn("dialect load error", ex);
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoaderEngine.java:63]: @tuichenchuxin Please use log.error.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoadMaterials.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    
+    private final Collection<LoadMaterial> loadMaterials;
+    
+    private final DatabaseType databaseType;
+    
+    @RequiredArgsConstructor
+    @Getter
+    public static class LoadMaterial {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/TableMetaDataLoadMaterials.java:40]: @tuichenchuxin Can we remove this inner class? And extract its filed to TableMetaDataLoadMaterials?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/util/TableMetaDataUtil.java [hunk 8]
```diff
@@ -36,13 +43,42 @@
...
+    
+    /**
+     * get table meta data load materials by one table.
+     *
+     * @param dataSource data source
+     * @param tableName table name
+     * @param databaseType database type
+     * @return TableMetaDataLoadMaterials
+     */
+    public static TableMetaDataLoadMaterials getTableMetaDataLoadMaterialsByOneTable(final DataSource dataSource, final String tableName, final DatabaseType databaseType) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/schema/builder/util/TableMetaDataUtil.java:77]: @tuichenchuxin Can these similar methods be combined into one method?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]

**Review by tristaZero — APPROVED** [review]

**Pull request merged**